// ng-gallery/esm2020/lib/components/adapters/common-adapter.mjs
import { DIRECTION_DOWN, DIRECTION_LEFT, DIRECTION_RIGHT, DIRECTION_UP } from './base-adapter';
/**
 * Common properties for both MainAdapter and ThumbAdapter
 */
export class HorizontalCommonAdapter {
    constructor(slider, config) {
        this.slider = slider;
        this.config = config;
        this.panDirection = DIRECTION_LEFT | DIRECTION_RIGHT;
        this.scrollSnapType = 'x mandatory';
    }
    get scrollValue() {
        return this.slider.scrollLeft;
    }
    get clientSize() {
        return this.slider.clientWidth;
    }
    getPanValue(value, e, behavior) {
        return {
            behavior,
            left: value - e.deltaX
        };
    }
}
export class VerticalCommonAdapter {
    constructor(slider, config) {
        this.slider = slider;
        this.config = config;
        this.scrollSnapType = 'y mandatory';
        this.panDirection = DIRECTION_UP | DIRECTION_DOWN;
    }
    get scrollValue() {
        return this.slider.scrollTop;
    }
    get clientSize() {
        return this.slider.clientHeight;
    }
    getPanValue(value, e, behavior) {
        return {
            behavior,
            top: value - e.deltaY
        };
    }
}
